java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -153,31 +158,59 @@ class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
      * If an id is provided, ensures that the id is registered to the [Store] of the [StoreRecord]
      */
     suspend fun getStore(muxId: String, id: Int? = null): StoreRecord<Data, Op, T> {
-        storeMutex.withLock {
+        stateMutex.withLock {
             val storeRecord = stores.getOrPut(muxId) {
                 setupStore(muxId)
             }
 
             // Register `id` with store if it is not already registered. It is necessary to register
-            // a callback per id because wehn teh Direct Store Muxer receives a message from the
+            // a callback per id because when the Direct Store Muxer receives a message from the
             // Direct Store, it utilizes the message id to determine which observer to redirect
             // the message to.
-            if (id != null && !storeRecord.idSet.contains(id)) {
-                synchronized(proxyManager) {
+            if (id != null) {
+                // Ensure 'id' has already registered with proxyManager
+                check(proxyManager.callbacks.containsKey(id)) {
+                    "Callback token is not registered to the Direct Store Muxer."
+                }
+
+                if (!storeRecord.idSet.contains(id)) {
                     storeRecord.store.on(
                         ProxyCallback { message ->
-                            proxyManager.getCallback(id)?.invoke(message.withMuxId(muxId))
+                            stateMutex.withLock { proxyManager.getCallback(id)?.invoke(message.withMuxId(muxId)) }
                         },
                         id
                     )
                     storeRecord.idSet.add(id)
                     callbackIdToMuxIdMap.getOrPut(id, { mutableSetOf() }).add(muxId)
                 }
             }
+
             return storeRecord
         }
     }
 
+    suspend fun consistentState(): Boolean = stateMutex.withLock {
```

Comment:
this is used only in tests. You can put the function in the test file, as an extension function
suspend fun DirectStoreMuxer.consistentState()